src/VS/Microsoft.Diagnostics.SOSPackage/SOSPackage.cs

```diff
@@ -0,0 +1,325 @@
+            if (pvaIn == IntPtr.Zero) {
+                // No arguments.
+                return VSConstants.E_INVALIDARG;
+            }
+
+            object vaInObject = Marshal.GetObjectForNativeVariant(pvaIn);
+            if (vaInObject == null || vaInObject.GetType() != typeof(string)) {
+                return VSConstants.E_INVALIDARG;
+            }
+
+            arguments = vaInObject as string;
+            return VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// Used to determine if the shell is querying for the parameter list.
+        /// </summary>
+        private static bool IsQueryParameterList(System.IntPtr pvaIn, System.IntPtr pvaOut, uint nCmdexecopt)
+        {
+            ushort lo = (ushort)(nCmdexecopt & (uint)0xffff);
+            ushort hi = (ushort)(nCmdexecopt >> 16);
+            if (lo == (ushort)OLECMDEXECOPT.OLECMDEXECOPT_SHOWHELP)
+            {
+                if (hi == VsMenus.VSCmdOptQueryParameterList)
+                {
+                    if (pvaOut != IntPtr.Zero) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        #endregion
+
+        #region VisualStudio.OLE.Interop.IServiceProvider
+
+        int VisualStudio.OLE.Interop.IServiceProvider.QueryService(ref Guid guidService, ref Guid riid, out IntPtr ppvObject)
+        {
+            if (guidService == SOSConcordService.guidSOSPackageService)
+            {
+                IntPtr unk = Marshal.GetIUnknownForObject(this);
+                int hr = Marshal.QueryInterface(unk, ref riid, out ppvObject);
+                Marshal.Release(unk);
+                return hr;
+            }
+            ppvObject = IntPtr.Zero;
+            return VSConstants.E_NOINTERFACE;
+        }
+
+        #endregion
+
+        #region IVsCustomDebuggerEventHandler110
+
+        int IVsCustomDebuggerEventHandler110.OnCustomDebugEvent(ref Guid processId, VsComponentMessage message)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            switch ((SOSPackageMessageId)message.MessageCode)
+            {
+                case SOSPackageMessageId.OutputStringNormal:
+                case SOSPackageMessageId.OutputStringWarning:
+                case SOSPackageMessageId.OutputStringError:
+                    if (message.Parameter1 is string text)
+                    {
+                        _commandWindow.OutputString(text);
+                        return VSConstants.S_OK;
+                    }
+                    break;
+            }
+
+            return VSConstants.E_INVALIDARG;
+        }
+
+        #endregion
+
+        private void OnContextChanged(Process newProcess, Program newProgram, EnvDTE.Thread newThread, EnvDTE.StackFrame newStackFrame)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            if (newProcess != null)
+            {
+                foreach (DkmProcess dkmProcess in DkmProcess.GetProcesses())
```

Comment:
>foreach [](http://example.com/codeflow?start=16&length=7)

Do you want to send a different message for each process when a context changes?

Reply:
As long as the same instance of my IDkmCustomMessageForwardReceiver's implementation class (SOSConcordServices) gets called, sending the set context message on only the first process makes sense.